Add a CGALExports.cmake file

bo0ts opened the pull request:
While looking into quick workarounds for issue #47 I realized that the only way to properly work around the problem is to add imported targets (reading LOCATION from an IMPORTED target is still fine) .

This also lays the foundation to compute the rest of the Config files from the targets and finally getting rid of two different configs.

The targets should remain undocumented for now since they do not have the proper include directories or compilation options and they still link too many not needed libraries directly.

lrineau commented:
Which CMake versions will support the code after this pull-request?

lrineau commented on Installation/cmake/modules/CGALConfig_install.cmake.in:
```diff
@@ -41,11 +44,19 @@ set(WITH_CGAL_ImageIO "@WITH_CGAL_ImageIO@" )
+CGAL_get_property_optional(CGAL_Core_LIBRARY TARGET CGAL::CGAL_Core PROPERTY LOCATION)
+CGAL_get_property_optional(CGAL_ImageIO_LIBRARY TARGET CGAL::CGAL_ImageIO PROPERTY LOCATION)
+CGAL_get_property_optional(CGAL_Qt3_LIBRARY TARGET CGAL::CGAL_Qt3 PROPERTY LOCATION)
+CGAL_get_property_optional(CGAL_Qt4_LIBRARY TARGET CGAL::CGAL_Qt4 PROPERTY LOCATION)
```
I am not sure, but it seems lines 55-59 will create a syntax error: the keywords `TARGET` and `PROPERTY` should probably be removed.

bo0ts commented:
Thanks, I fixex and rebased the bug you reported. I changed the file in it's intalled location when I noticed the mistake instead of changing the file used to generate it.

I did preliminary tests on Debian-stable (cmake 3.0) and CentOS 5 (cmake 2.6) and both seemed OK.

bo0ts commented:
I also just realized that this will break when auto-linking is enabled. I'll add a patch for this, meanwhile please keep this open.

bo0ts commented:
The recent commits should fix any possible issues with auto-linking on MSVC. OK, to merge into integration.

lrineau commented:
I cannot do it now. If you are in front of your PC, please merge the pr or the corresponding branch in integration. Follow steps from the wiki if you have doubts on the method. Remember integration does not fast-forward. You need to reset your local copy.

lrineau commented:
Merged in `integration`. Will be tested in [CGAL-4.7-Ic-69](https://cgal.geometryfactory.com/CGAL/Members/testsuite/results-4.7-Ic-69.shtml), on Monday night.

lrineau commented:
This pull-request was the only one modifying the CMake scripts of CGAL libraries, in [CGAL-4.7-Ic-72][Ic-72], and all Apple results have issues with the CGAL libraries (but `libCGAL` itself). It seems that, during the link step of CGAL libraries, `libCGAL` is no longer linked with. See [the configuration/compilation log of the first one `x86-64_Darwin-13.0_Apple-clang-3.2_Debug`][log].

[Ic-72]: https://cgal.geometryfactory.com/CGAL/Members/testsuite/results-4.7-Ic-72.shtml
[log]: https://cgal.geometryfactory.com/CGAL/Members/testsuite/CGAL-4.7-Ic-72/Installation/TestReport_cgaltester_x86-64_Darwin-13.0_Apple-clang-3.2_Debug.gz

bo0ts commented:
The issue seems to be the use of the `CGAL_LIBRARY` variable in the CMakeLists.txt of `CGAL_Core`, `CGAL_ImageIO`, and `CGAL_Qt3`. This variable is not defined anymore within the build-system because it should be replaced by the target `CGAL`.

This bug can be triggered on a Linux machine by passing `-Wl,--no-undefined` to `CMAKE_SHARED_LINKER_FLAGS`.

I'll add a commit that links those libraries to `libCGAL` using the `CGAL` target. Actually, those libraries should only link to `libCGAL` in a release build, but I'm afraid that change would break too much since some users might rely on the fact that `libCGAL_ImageIO` etc. always carry `libCGAL` as a dependency.

lrineau commented:
With this PR merged in `master`, CMake-3.0.2 warns about [`CMP0023`](http://www.cmake.org/cmake/help/v3.0/policy/CMP0024.html), if a branch-build is configured with `-DWITH_demos:BOOL=ON` or `-DWITH_examles:BOOL=ON`:

```
-- Configuring examples in examples/AABB_tree
CMake Warning (dev) at build/CGALConfig.cmake:12 (include):
  Policy CMP0024 is not set: Disallow include export result.  Run "cmake
  --help-policy CMP0024" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  The file

    /home/lrineau/Git/cgal/build/CGALExports.cmake

  was generated by the export() command.  It should not be used as the
  argument to the include() command.  Use ALIAS targets instead to refer to
  targets by alternative names.

Call Stack (most recent call first):
  Installation/cmake/modules/FindCGAL.cmake:77 (include)
  build/examples/AABB_tree/CMakeLists.txt:16 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.
```

bo0ts commented:
This is the result of calling `find_package(CGAL)` from within the CGAL build tree. This will lead to inclusion of `CGALExports.cmake` via `CGALConfig.cmake`. 

There are several options to fix this:

- `FindCGAL.cmake` should know when it is called from within a CGAL build and not use the config
- Demo/Example `CMakeLists.txt` should know when they are within a CGAL build tree

The second option would technically be the correct option, but it will not work because we rely on variables set in CGALConfig.cmake and not only on imported targets.

lrineau commented:
When you say `FindCGAL.cmake`, are you talking about `CGALConfig.cmake`? Because even if there is a file named `FindCGAL.cmake` in `Installation/cmake/modules/`, it is not used. The CMake API of CGAL uses the *Config* mode of `find_package`

lrineau commented:
If think the `CGALConfig.cmake` file could start with:
```
if(TARGET CGAL)
  # We are in a branch-build. Do nothing.
  return()
endif()
```

lrineau commented:
The comment about the branch-build is wrong. It could be a full build of CGAL, from a CGAL tarball.

bo0ts commented:
> When you say FindCGAL.cmake, are you talking about CGALConfig.cmake? Because even if there > is a file named FindCGAL.cmake in Installation/cmake/modules/, it is not used. The CMake API of CGAL uses the Config mode of find_package

You are correct that `FindCGAL.cmake` is not invoked normally, but here it is since `FindCGAL.cmake` is in our adjusted module path.

> if(TARGET CGAL)
>   # We are in a branch-build. Do nothing.
>   return()
> endif()

This approach is too naive. The detection might be wrong (CGAL could
be a user target), but we already have the variable
`CGAL_BUILDING_LIBS` for this purpose. It also is not correct to just
do nothing since the `CGALConfig.cmake` now contains variables (like
`CGAL_LIBRARY`) which are deduced from imported targets.

I'll add a commit which adds the appropriate alias targets.

lrineau commented:
What about CMP0026?

bo0ts commented:
This PR also fixes the conflict with CMP0026, we still read the `LOCATION` property but only from imported targets (not build targets), which works fine: an imported target is complete at the time of its import.

lrineau commented:
I have merged this PR #133 in the PR #154. Probably #133 will be merged in `master` through #154.

lrineau commented:
The `ALIAS` feature is new in CMake-2.8.12! My proposal: e43b1e4 from PR #171.

bo0ts commented:
How do you want to proceed? I can either incorporate your changes in this PR or we can merge it through #171 or #154 .

lrineau commented:
I prefer that you merge or cherry pick my change in your PR, so that it is self-contained.

bo0ts commented:
I think the necessary commits are  

- @lrineau 	Check if the libraries are actually defined … 	fb4975f
- @lrineau 	Avoid multiple definitions of the target aliases … 	877634b
- @lrineau 	Do include the export file if CMP0024 does not exist … 	e43b1e4

am I missing any?
